jans-cedarling/cedarling/src/jwt/service.rs

```diff
@@ -0,0 +1,130 @@
+/*
+ * This software is available under the Apache-2.0 license.
+ * See https://www.apache.org/licenses/LICENSE-2.0.txt for full text.
+ *
+ * Copyright (c) 2024, Gluu, Inc.
+ */
+
+use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, Header, Validation};
+use std::sync::Arc;
+
+use crate::{models::token_data::TokenPayload, JwtConfig};
+
+use super::{key_service::KeyService, Error};
+
+/// Service for JWT validation.
+///
+/// The `JwtService` is responsible for decoding and validating JSON Web Tokens (JWTs).
+/// It can operate in two modes: with validation, using a specified key service for key retrieval,
+/// and without validation, for cases where the token's validity is not required.
+#[derive(Clone)]
+pub struct JwtService {
+    config: Arc<JwtConfig>,
+    key_service: Option<Arc<dyn KeyService>>,
+}
+
+impl JwtService {
+    /// Creates a new instance of `JwtService`.
+    ///
+    /// If `CEDARLING_JWT_VALIDATION` is set to `Enabled` in the
+    /// bootstrap properties `key_service` cannot not be `None`.
+    pub fn new(config: JwtConfig, key_service: Option<Arc<dyn KeyService>>) -> Self {
+        Self {
+            config: Arc::new(config),
+            key_service,
+        }
+    }
+
+    /// Decodes a JWT, optionally validating it based on the configuration settings.
+    pub fn decode<T: serde::de::DeserializeOwned>(&self, jwt: &str) -> Result<T, Error> {
+        match self.config.as_ref() {
+            JwtConfig::Disabled => decode_jwt_without_validation(jwt),
+            JwtConfig::Enabled {
+                signature_algorithms,
+            } => self.decode_and_validate_jwt(jwt, signature_algorithms),
+        }
+    }
+
+    /// Decodes a JWT and returns the `TokenPayload`.
+    pub fn decode_token_data(&self, jwt: &str) -> Result<TokenPayload, Error> {
+        self.decode(jwt)
+    }
+
+    /// Decodes and validates a JWT using the specified algorithms and keys.
+    fn decode_and_validate_jwt<T: serde::de::DeserializeOwned>(
+        &self,
+        jwt: &str,
+        signature_algorithms: &Vec<String>,
+    ) -> Result<T, Error> {
+        let header = extract_jwt_header(jwt).map_err(|e| Error::ParsingError(e))?;
+
```

Comment:
we can just pass the function

```suggestion
        let header = extract_jwt_header(jwt).map_err(Error::ParsingError)?;
```

Reply:
this is fixed. that function no longer exists.